File: 백준/Gold/23309. 철도 공사/철도 공사.java
import java.io.*;
import java.util.StringTokenizer;
public class Main {
    static int[] pre;
    static int[] nxt;
    static int[] stations;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        pre = new int[1000001];
        nxt = new int[1000001];
        stations = new int[N];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            stations[i] = Integer.parseInt(st.nextToken());
        }

        for (int i = 0; i < N; i++) {
            int now = stations[i];
            pre[now] = stations[(i+N-1) % N];
            nxt[now] = stations[(i+1) % N];
        }

        StringBuilder sb = new StringBuilder();
        for (int p = 0; p < M; p++) {
            st = new StringTokenizer(br.readLine());
            String command = st.nextToken();
            if (command.equals("BN")) {
                int i = Integer.parseInt(st.nextToken());
                int j = Integer.parseInt(st.nextToken());
                sb.append(String.valueOf(nxt[i])).append("\n");
                insert(i, j);
            } else if (command.equals("BP")) {
                int i = Integer.parseInt(st.nextToken());
                int j = Integer.parseInt(st.nextToken());
                sb.append(String.valueOf(pre[i])).append("\n");
                insert(pre[i], j);
            } else if (command.equals("CN")) {
                int i = Integer.parseInt(st.nextToken());
                sb.append(String.valueOf(nxt[i])).append("\n");
                remove(nxt[i]);
            } else {
                int i = Integer.parseInt(st.nextToken());
                sb.append(String.valueOf(pre[i])).append("\n");
                remove(pre[i]);
            }
        }
        System.out.println(sb);
    }

    // A > B > C -> A > B > D > C
    static void insert(int B, int D) {
        int C = nxt[B];
        nxt[B] = D;
        pre[D] = B;
        nxt[D] = C;
        pre[C] = D;
    }

    // A > B > C -> A  > C
    static void remove(int B) {
        int A = pre[B];
        int C = nxt[B];
        nxt[A] = C;
        pre[C] = A;
    }
}

File: 백준/Gold/23309. 철도 공사/철도 공사.py
import sys
import os, io, __pypy__
input = io.BytesIO(os.read(0, os.fstat(0).st_size)).readline
ans = __pypy__.builders.StringBuilder()
def main():
    N, M = map(int, input().split())
    stations_list = list(map(int, input().split()))

    prv = [0 for _ in range(1_000_000 + 1)]
    nxt = [0 for _ in range(1_000_000 + 1)]

    for i in range(N):
        now_station_num = stations_list[i]
        prv[now_station_num] = stations_list[i-1]
        nxt[now_station_num] = stations_list[(i + 1) % N]

    # A > B > C -> A > B > D > C
    def insert(B, D):
        C = nxt[B]
        nxt[B] = D
        prv[D], nxt[D] = B, C
        prv[C] = D

    # A > B > C -> A > C
    def remove(B):
        A, C = prv[B], nxt[B]
        nxt[A] = C
        prv[C] = A

    for _ in range(M):
        command = input().split()
        if command[0] == b'BN':
            i, j = int(command[1]), int(command[2])
            ans.append(str(nxt[i]) + '\n')
            insert(i, j)
        elif command[0] == b'BP':
            i, j = int(command[1]), int(command[2])
            ans.append(str(prv[i]) + '\n')
            insert(prv[i], j)
        elif command[0] == b'CN':
            i = int(command[1])
            ans.append(str(nxt[i]) + '\n')
            remove(nxt[i])
        else:
            i = int(command[1])
            ans.append(str(prv[i]) + '\n')
            remove(prv[i])

    os.write(1, ans.build().encode())


if __name__ == '__main__':
    main()